### PR TITLE
alien language

### PR DESCRIPTION
**Pull request by wliu015**

### CODE REVIEW

File: alien_langurage.cpp [hunk 1]
```diff
@@ -0,0 +1,112 @@
+// A C++ program to order of characters in an alien language
+#include <iostream>
+#include <unordered_map>
+#include <vector>
+using namespace std;
+
+
```
Comment (wliu015) [inline alien_langurage.cpp]: test

File: alien_langurage.cpp [hunk 2]
```diff
@@ -0,0 +1,112 @@
...
+class Graph {
+public:
+    Graph();
+    // add an edge to graph
+    void AddEdge(char v, char w);
+    // prints a Topological Sort of the complete graph
```
Comment (wliu015) [inline alien_langurage.cpp]: WTF

File: alien_langurage.cpp [hunk 3]
```diff
@@ -0,0 +1,112 @@
...
+}
+
+void Graph::AddEdge(char v, char w) {
+    V_[v] = true;
+    V_[w] = true;
+    adj_[v].push_back(w); // Add w to v’s list.
+
+}
+
+void Graph::DFS(char v, vector<Status> &visit, vector<char> &result) {
```
Comment (qiqjiao) [inline alien_langurage.cpp]: return boolean to indicate whether the graph is DAG or has circle

File: alien_langurage.cpp [hunk 4]
```diff
@@ -0,0 +1,112 @@
...
+    // Mark the current node as visiting.
+    visit[v] = kVisiting;
+    // Recur for all the vertices adjacent to this vertex
+    for (char u : adj_[v]) {
+        if (visit[u]==kUnvisited)    DFS(u, visit, result);
```
Comment (qiqjiao) [inline alien_langurage.cpp]: add space around operator (e.g. ==) here and elsewhere

File: alien_langurage.cpp [hunk 5]
```diff
@@ -0,0 +1,112 @@
...
+            return;
+        }
+    }
+    visit[v] = kVisited;
+    //push the visited node to the result
+    result.push_back(v);
+}
+
+void Graph::TopologicalSort() {
+    vector<char> result;
```
Comment (qiqjiao) [inline alien_langurage.cpp]: std::string is more appropriate to represent array of characters

File: alien_langurage.cpp [hunk 6]
```diff
@@ -0,0 +1,112 @@
...
+    for(const auto & kv : adj_){
+        cout << kv.first << ":\t" ;
+        for(char ch : kv.second)    cout << ch << " ";
+        cout << endl;
+    }
+}
+// This function fidns and prints order of characer from a sorted
+// array of words.
+void PrintOrder(const vector<string> &words)  {
+    int n = words.size();
```
Comment (qiqjiao) [inline alien_langurage.cpp]: const int

File: alien_langurage.cpp [hunk 7]
```diff
@@ -0,0 +1,112 @@
...
+    Graph graph;
+    // Process all adjacent pairs of words and create a graph
+    for (int i = 0; i < n-1; i++) {
+        // Take the current two words and find the first mismatching
+        // character
+        string word1 = words[i], word2 = words[i+1];
```
Comment (qiqjiao) [inline alien_langurage.cpp]: const string &word1 = ...
const string &word2 = ...

File: alien_langurage.cpp [hunk 8]
```diff
@@ -0,0 +1,112 @@
...
+        for (int j = 0; j < min(word1.length(), word2.length()); j++){
```
Comment (qiqjiao) [inline alien_langurage.cpp]: you need to visit every character of each word. for this case {"ax", "by"}, 'x', 'y' are not considered.

File: alien_langurage.cpp (hunk 3 above)
Comment (wliu015) [reply alien_langurage.cpp]: Good point, Thanks.

File: alien_langurage.cpp (hunk 4 above)
Comment (wliu015) [reply alien_langurage.cpp]: Yes, right.

File: alien_langurage.cpp (hunk 5 above)
Comment (wliu015) [reply alien_langurage.cpp]: Good point. In C++11 string Copy-On-Write is banned. So string is better here.

File: alien_langurage.cpp (hunk 7 above)
Comment (wliu015) [reply alien_langurage.cpp]: Done

File: alien_langurage.cpp (hunk 8 above)
Comment (wliu015) [reply alien_langurage.cpp]: I don't think so. You need to find the first mismatching character only. in this case you can get  'a' > 'b', but you don't know the order between 'x' and 'y'

File: findMinSteps.cpp [hunk 9]
```diff
@@ -0,0 +1,79 @@
...
+        }
+        maze[i][j] = -1;
+    }
+}
+
+int findMinNumSteps(vector<vector<int>> &maze,
+                    int exitRow, int exitCol)
+{
+    int ret = -1;
+    int rows = maze.size();
```
Comment (qiqjiao) [inline findMinSteps.cpp]: 1. check maze.empty at the top
2. const int rows, cols

File: findMinSteps.cpp [hunk 10]
```diff
@@ -0,0 +1,79 @@
+#include <iostream>
+#include <vector>
+#include <queue>
+using namespace std;
+
+
+void BFS(vector<vector<int>> maze, int rows, int columns,
+         int startX, int startY,
+         int exitX, int exitY,
+         int *ret)
```
Comment (qiqjiao) [inline findMinSteps.cpp]: There's only 1 output, so just return it

File: findMinSteps.cpp [hunk 11]
```diff
@@ -0,0 +1,79 @@
...
+                maze[ti][tj]==1)
+            {
+                q.push(make_pair(ti, tj));
+            }
+        }
+        maze[i][j] = -1;
+    }
+}
+
+int findMinNumSteps(vector<vector<int>> &maze,
```
Comment (qiqjiao) [inline findMinSteps.cpp]: This is java style, change to FindMinNumSteps, exit_row, exit_col, etc.

File: findMinSteps.cpp [hunk 12]
```diff
@@ -0,0 +1,79 @@
+#include <iostream>
+#include <vector>
+#include <queue>
+using namespace std;
+
+
+void BFS(vector<vector<int>> maze, int rows, int columns,
```
Comment (qiqjiao) [inline findMinSteps.cpp]: Use a function name that indicates the functionality, like DoFindMinNumSteps

File: findMinSteps.cpp [hunk 13]
```diff
@@ -0,0 +1,79 @@
...
+        q.push(make_pair(startX, startY));
+        maze[startX][startY] = -1;
+    }
+
+    while(!q.empty()){
+        int i = q.front().first;
+        int j = q.front().second;
+        q.pop();
+        if(i == -1 && j == -1) {
+            // trick 1: terminate if queue just contain <-1, -1>
```
Comment (qiqjiao) [inline findMinSteps.cpp]: It takes time to understand the trick. I suggest redesign the maze data structure. e.g.
struct Cell {
  bool can_through;
  bool is_visited;
  int level;
};
using Maze = vector<vector<Cell>>;

In this way magic number '-1' can be avoided, also maintaining the level becomes straightforward.

File: findMinSteps.cpp [hunk 14]
```diff
@@ -0,0 +1,79 @@
...
+         int exitX, int exitY,
+         int *ret)
+{
+    queue<pair<int, int>> q;
+    q.push(make_pair(-1, -1));
+    const int kMOVES = 4; //size of moves
+    int moves[kMOVES][2] = {{0,1},{1,0},{0,-1},{-1,0}};
+    int level = -1;
+
+    if(maze[startX][startY]==1)   {
```
Comment (qiqjiao) [inline findMinSteps.cpp]: space after 'if', 'while', and other keywords.
space around operator '=='

File: findMinSteps.cpp [hunk 15]
```diff
@@ -0,0 +1,79 @@
...
+            if(q.empty()) return;
+            ++level;
+            q.push(make_pair(-1, -1));
+            continue;
+        }
+        if(i == exitX && j == exitY) {
+            *ret = level;
+            return;
+        }
+        for (int l = 0;l < kMOVES;++l){
```
Comment (qiqjiao) [inline findMinSteps.cpp]: space after ';'

File: findMinSteps.cpp (hunk 9 above)
Comment (wliu015) [reply findMinSteps.cpp]: Yes. Gonna make the  maze const. And check the maze.empty() first.

File: findMinSteps.cpp (hunk 10 above)
Comment (wliu015) [reply findMinSteps.cpp]: Right.

File: findMinSteps.cpp (hunk 9 above)
Comment (wliu015) [reply findMinSteps.cpp]: Genius. Code becomes clean this way.

File: findMinSteps.cpp [hunk 16]
```diff
@@ -0,0 +1,103 @@
...
+            {1, 0, 1, 0, 0}  
+    };  
+    cout << FindMinNumSteps(maze1, 4, 3) << endl; //7  
+    cout << FindMinNumSteps(maze1, 1, 1) << endl; //2  
+    cout << FindMinNumSteps(maze1, 0, 0) << endl; //0  
+    cout << FindMinNumSteps(maze1, 2, 0) << endl; //-1  
+    cout << FindMinNumSteps(maze1, 5, 2) << endl; //7  
+    Maze maze2;  
+    cout << FindMinNumSteps(maze2, 5, 2) << endl; //-1  
+
```
Comment (qiqjiao) [inline findMinSteps.cpp:101]: Looks really good and clean!

I suggest adding following case to test the 0 starting point:
  {0, 1},
  {1, 1}

**Pull request merged**